source/common/router/rds_impl.h

```diff
@@ -60,121 +64,65 @@ class RouteConfigProviderUtil {
       const std::string& stat_prefix, RouteConfigProviderManager& route_config_provider_manager);
 };
 
-class RouteConfigProviderManagerImpl;
-
 /**
  * Implementation of RouteConfigProvider that holds a static route configuration.
  */
 class StaticRouteConfigProviderImpl : public RouteConfigProvider {
 public:
   StaticRouteConfigProviderImpl(const envoy::config::route::v3::RouteConfiguration& config,
-                                const OptionalHttpFilters& http_filters,
+                                Rds::ConfigTraits& config_traits,
                                 Server::Configuration::ServerFactoryContext& factory_context,
-                                ProtobufMessage::ValidationVisitor& validator,
-                                RouteConfigProviderManagerImpl& route_config_provider_manager);
+                                Rds::RouteConfigProviderManager& route_config_provider_manager);
   ~StaticRouteConfigProviderImpl() override;
 
   // Router::RouteConfigProvider
-  Router::ConfigConstSharedPtr config() override { return config_; }
-  absl::optional<ConfigInfo> configInfo() const override {
-    return ConfigInfo{route_config_proto_, ""};
-  }
-  SystemTime lastUpdated() const override { return last_updated_; }
-  void onConfigUpdate() override {}
+  Rds::ConfigConstSharedPtr config() override { return base_.config(); }
+  absl::optional<ConfigInfo> configInfo() const override { return base_.configInfo(); }
+  SystemTime lastUpdated() const override { return base_.lastUpdated(); }
+  void onConfigUpdate() override { base_.onConfigUpdate(); }
+  ConfigConstSharedPtr configCast() override;
   void requestVirtualHostsUpdate(const std::string&, Event::Dispatcher&,
                                  std::weak_ptr<Http::RouteConfigUpdatedCallback>) override {
     NOT_IMPLEMENTED_GCOVR_EXCL_LINE;
   }
 
 private:
-  ConfigConstSharedPtr config_;
-  envoy::config::route::v3::RouteConfiguration route_config_proto_;
-  SystemTime last_updated_;
-  RouteConfigProviderManagerImpl& route_config_provider_manager_;
-};
-
-/**
- * All RDS stats. @see stats_macros.h
- */
-#define ALL_RDS_STATS(COUNTER, GAUGE)                                                              \
-  COUNTER(config_reload)                                                                           \
-  COUNTER(update_empty)                                                                            \
-  GAUGE(config_reload_time_ms, NeverImport)
-
-/**
- * Struct definition for all RDS stats. @see stats_macros.h
- */
-struct RdsStats {
-  ALL_RDS_STATS(GENERATE_COUNTER_STRUCT, GENERATE_GAUGE_STRUCT)
+  Rds::StaticRouteConfigProviderImpl base_;
```

Comment:
Why is `Rds::StaticRouteConfigProviderImpl` a composed field (`base_`) instead of deriving the class from `Rds::StaticRouteConfigProviderImpl`?

Reply:
The `Router::RouteConfigProvider` adds couple methods to the `Rds::RouteConfigProvider` interface.
If I inherit from `Rds::StaticRouteConfigProviderImpl` those method are lost. However for static provider they are mostly not called, but called for dynamic providers. And static and dynamic providers should implement the same interface.